src/transformers/trainer.py

```diff
@@ -595,7 +595,8 @@ def __init__(
             if args.half_precision_backend == "cuda_amp":
                 self.use_cuda_amp = True
                 self.amp_dtype = torch.float16 if args.fp16 else torch.bfloat16
-                self.do_grad_scaling = True
+                #  bf16 does not need grad scaling
+                self.do_grad_scaling = self.amp_dtype == torch.float16
```

Comment:
Yes to this, and I think we can put the big test below under `if self.do_grad_scaling`.

Reply:
thanks! added